### PR TITLE
8279545: Buffer overrun in reverse_words of sharedRuntime_x86_64.cpp:3517

### PR DESCRIPTION
**Pull request by hseigel**
Please review this small fix to prevent a potential integer overflow that can cause a buffer overrun.  The fix was tested with Mach5 tiers 1-2 on Linux, Mac OS, and Windows, Mach5 tiers 3-5 on Linux x64, and Mach5 builds of the JDK on PPC and S390.

Thanks, Harold

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279545](https://bugs.openjdk.java.net/browse/JDK-8279545): Buffer overrun in reverse_words of sharedRuntime_x86_64.cpp:3517

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)
 * [Lois Foltan](https://openjdk.java.net/census#lfoltan) (@lfoltan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7053/head:pull/7053` \
`$ git checkout pull/7053`

Update a local copy of the PR: \
`$ git checkout pull/7053` \
`$ git pull https://git.openjdk.java.net/jdk pull/7053/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7053`

View PR using the GUI difftool: \
`$ git pr show -t 7053`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7053.diff">https://git.openjdk.java.net/jdk/pull/7053.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp [hunk 1]
```diff
@@ -3246,8 +3246,8 @@ void SharedRuntime::montgomery_square(jint *a_ints, jint *n_ints,
   // Make very sure we don't use so much space that the stack might
   // overflow. 512 jints corresponds to an 16384-bit integer and
   // will use here a total of 6k bytes of stack space.
+  guarantee(longwords <= 341, "must be");
```
Comment (coleenp) [inline src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:3249]: I was mystified about this number.  Can you make this instead like:

int thing = sizeof (unsigned long) * 3;
guarantee(longwords <= 8192 / thing, "must be");

Or is division just as unsafe?

Should this be uint64_t, not long ?  Or change it to julong if it's a small change to match the x86 version.

Comment (hseigel) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:3249]: How about just:    guarantee(longwords <= 8092 / (sizeof(julong) * 3), "must be");   ?
I don't think that doing division is a problem.

**Review by coleenp — APPROVED** [review]
Looks good, thanks!

**Review by lfoltan — APPROVED** [review]
Looks good.
Lois

**Comment by hseigel** [comment]
Thanks Coleen and Lois for the reviews!

/integrate

**Pull request closed without merging**